### DISCUSSION TITLE
Improvements and Requests

### DISCUSSION THREAD

**Discussion by jmcdo29** [jmcdo29/ogma#527] [Ideas]
Hey everyone, so I've found a bug that I didn't originally think about and you can read about it at #526. For the fix, I'm thinking of printing the error inline the way `process.stdout.write()` would

```
[2021-03-16T15:41:33.207Z] [INFO]  [jay-pop] [demo] 538029 Error: This is an error
```

The other option would be to fully serialize the error object and print it similar to how `console.log()` does

```
[2021-03-16T15:41:33.207Z] [INFO]  [jay-pop] [demo] 538029
Error: This is an error
    at REPL45:1:13
    at Script.runInThisContext (vm.js:133:18)
    at REPLServer.defaultEval (repl.js:484:29)
    at bound (domain.js:413:15)
    at REPLServer.runBound [as eval] (domain.js:424:12)
    at REPLServer.onLine (repl.js:817:10)
    at REPLServer.emit (events.js:327:22)
    at REPLServer.EventEmitter.emit (domain.js:467:12)
    at REPLServer.Interface._onLine (readline.js:337:10)
    at REPLServer.Interface._line (readline.js:666:8)
```
starting the log on a new line. 

This only effects string mode, but I wanted to try and get some feedback before committing to one over the other. I'm in favor of the former, but I'd love to hear the community's opinion too.

**Comment by jonatasfernandespimenta** [comment]
I really like the first option, I prefer having cleaner logs. Maybe an option to choose between the full log and the cleaner log would be a good idea, something like
```js
      service: {
        color: true,
        json: false,
        logLevel: 'ALL',
        application: 'NestJS',
        fullError: true
      },
```

**Reply by jmcdo29** [reply]
I was actually thinking of something similar for those who wanted to print the stack trace. Ogma does have a `printError` method which breaks up the error, but the problem arises when using Nest's `Logger` interface, as that property doesn't exist.

I'll probably add it in under a flag though, and if the flag s there then print the stack trace the way `console.log` does under the original log line and indented

**Reply by tukusejssirs** [reply]
@jmcdo29, is there a way to output stack traces using `@ogma/nestjs-module`? I’d like to see the stack traces, however, I [couldn’t find]https://github.com/jmcdo29/ogma/tree/main/packages/logger#ogma-options) a `fullError` (or similar option. Did you not implemented it yet? 🤔 Or if it is possible, how should I configure it?


Thanks! 🙏

**Reply by jmcdo29** [reply]
@tukusejssirs if I recall, I ended up just fixing the bug and leaving the use of `printError`. However, you can pass a stack trace as the second parameter of `error`, for comparison with nest's logger

**Reply by tukusejssirs** [reply]
> However, you can pass a stack trace as the second parameter of error, for comparison with nest's logger

You mean _manually_ passing the stack trace? I’d expect an _automatic_ stack trace creation (just like in `console.log(new Error('Some error')`). If I can pass a second argument like in `allowStackTrace: true`, that is okay with me, but I don’t want to create the stack trace by hand. Maybe I misunderstand you comment. :thinking:

> the second parameter of error

Do you mean the second parameter of `new Error()`? It seems like that accepts only a single `string` parameter. Again, maybe I miss something.

**Reply by jmcdo29** [reply]
> but I don’t want to create the stack trace by hand

All `Error` instances have a `.stack` property that contain their stack trace, so you can do `logger.error(error.name, error.stack)`

Also, if you do `logger.printError(error)` then you'll get three automated logs at different levels. The error's name at the `error` level, the error's message at the `warn` level and the error's stack trace at the `verbose` level
